### PR TITLE
Fix empty key in slice

### PR DESCRIPTION
**Pull request by alpe**
This patch fixes an issue with secondary index keys that are nil or empty. They are ignored now.

* Added tests for indexers

### CODE REVIEW

File: incubator/orm/indexer.go [hunk 1]
```diff
@@ -148,3 +148,19 @@ func makeIndexPrefixScanKey(indexKey []byte, rowID uint64) []byte {
...
+// pruneEmptyKeys drops any empty key from IndexerFunc f returned
+func pruneEmptyKeys(f IndexerFunc) IndexerFunc {
+	return func(v interface{}) ([][]byte, error) {
+		keys, err := f(v)
+		if err != nil {
+			return nil, err
+		}
+		for i := 0; i < len(keys); i++ {
+			if len(keys[i]) == 0 {
+				keys = append(keys[0:i], keys[i+1:]...)
```
Comment (aaronc) [inline incubator/orm/indexer.go]: I'm a bit wary of these types of operations in general. I'm assuming `append` doesn't allocate here because the slice is getting smaller if anything. The thing that seems potentially problematic is that it seems like you're mutating a slice while iterating over it. So whenever you hit an empty key, the `i+1`th element now becomes the `i`th element, but since you've already checked the `i`th element, I think the `i+1`th element wouldn't actually get checked. Am I right about that or am I missing something?

Also, I don't generally like to call `len` in the test condition of a for loop as it will get evaluated everything, here likely zero cost but I'm remembering `strlen` from C.

Comment (alpe) [reply incubator/orm/indexer.go]: np, I have redesigned the func.

**Review by aaronc — APPROVED** [review]
This definitely looks safer and am fine with this approach.

Do you think it's maybe worth trying a 0-allocation variant like you were trying to achieve or not worth the risk?

**Comment by alpe** [comment]
> This definitely looks safer and am fine with this approach.
> 
> Do you think it's maybe worth trying a 0-allocation variant like you were trying to achieve or not worth the risk?

I would expect the number of secondary index keys for an object to be small. Bigger number would run into Gas limits anyway. Optimizations should come with benchmarks. I would prefer to keep this a bugfix patch and have a new task for benchmarks.

**Comment by alpe** [comment]
I have added #12 as a reminder for benchmarks.

**Comment by aaronc** [comment]
> I have added #12 as a reminder for benchmarks.

Thanks @alpe
